core_data_modules/traced_data/io.py

```diff
@@ -16,16 +17,94 @@
 _td_type_error_string = "argument 'data' contains an element which is not of type TracedData"
 
 
+class _TracedDataIOUtil(object):
+    """
+    A collection of utility functions used by the IO classes in this file.
+    
+    Not for external use.
+    """
+
+    @staticmethod
+    def unique_messages(data, key_of_raw):
+        """
+        Filters a collection of TracedData objects such that there is only one object with each value for the given key.
+        
+        If there are multiple TracedData objects with the same value for the given key, one of those will be selected
+        arbitrarily; the rest will be dropped.
+        
+        :param data: TracedData objects to select unique messages from.
+        :type data: iterable of TracedData
+        :param key_of_raw: Key in TracedData objects of raw messages which should be unique in output.
+        :type key_of_raw: str
+        :return: TracedData objects distinct by the given key.
+        :rtype: list of TracedData
+        """
+        seen = set()
+        unique_data = []
+        for td in data:
+            if not td[key_of_raw] in seen:
+                seen.add(td[key_of_raw])
+                unique_data.append(td)
+                
+        return unique_data
+
+    @staticmethod
+    def unique_messages_with_code(data, key_of_raw, key_of_coded):
```

Comment:
I recommend separating the filtration part (selecting the unique values) from the verification part (raising assert exceptions if they're not unique)